crates/sui-data-ingestion-core/src/lib.rs

```diff
@@ -23,19 +24,19 @@ pub use worker_pool::WorkerPool;
 
 #[async_trait]
 pub trait Worker: Send + Sync {
-    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<()>;
-    /// Optional method. Allows controlling when workflow progress is updated in the progress store.
-    /// For instance, some pipelines may benefit from aggregating checkpoints, thus skipping
-    /// the saving of updates for intermediate checkpoints.
-    /// The default implementation is to update the progress store for every processed checkpoint.
-    async fn save_progress(
-        &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        Some(sequence_number)
-    }
+    type Result: Send + Sync;
+    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<Self::Result>;
 
     fn preprocess_hook(&self, _: &CheckpointData) -> Result<()> {
         Ok(())
     }
 }
+
+#[async_trait]
+pub trait Reducer<R: Send + Sync>: Send {
+    async fn commit(&self, batch: Vec<R>) -> Result<CheckpointSequenceNumber>;
```

Comment:
I don't quite understand this interface. Why does it need to return a checkpoint number. Also it is not necessarily always the case that `commit` is able to derive the checkpoint sequence number from `Vec<R>`.